tardis/simulation/base.py

```diff
@@ -401,6 +406,137 @@ def to_hdf(self, model, path_or_buf, path='', plasma_properties=None):
         self.runner.to_hdf(path_or_buf, path)
         model.to_hdf(path_or_buf, path, plasma_properties)
 
+    def make_source_function(self, model):
+        """
+        Calculates the source function using the line absorption rate estimator `Edotlu_estimator`
+
+        Formally it calculates the expresion ( 1 - exp(-tau_ul) ) S_ul but this product is what we need later,
+        so there is no need to factor out the source function explicitly.
+
+        Parameters
+        ----------
+        model : tardis.model.Radial1DModel
+
+        Returns
+        -------
+        DataFrame containing ( 1 - exp(-tau_ul) ) S_ul
+        """
+
+        upper_level_index = model.atom_data.lines.set_index(['atomic_number', 'ion_number', 'level_number_upper']).index.copy()
+        e_dot_lu = pd.DataFrame(self.runner.Edotlu, index=upper_level_index)
+        e_dot_u = e_dot_lu.groupby(level=[0, 1, 2]).sum()
+        e_dot_u.index.names = ['atomic_number', 'ion_number', 'source_level_number'] # To make the q_ul e_dot_u product work, could be cleaner
+        transitions = model.atom_data.macro_atom_data[model.atom_data.macro_atom_data.transition_type == -1].copy()
+        transitions_index = transitions.set_index(['atomic_number', 'ion_number', 'source_level_number']).index.copy()
+        tmp = model.plasma.transition_probabilities[(model.atom_data.macro_atom_data.transition_type == -1).values]
+        q_ul = tmp.set_index(transitions_index)
+        return (model.atom_data.lines.wavelength[transitions.transition_line_id].values.reshape(-1,1) * 
+                 (q_ul * e_dot_u) * model.tardis_config.supernova.time_explosion / (4*np.pi) )
+
+    def integrate(self,model):
+        num_shell, = self.runner.volume.shape
+        ps         = np.linspace(0.999, 0, num = 3 * num_shell)
```

Comment:
The outside in order leads to a situation where complexity increases as you step further, which I think is a simpler way to organize the work. For example, if you start from the outside, rays at the first impact parameter (the ps) will only intersect the outer shell, while at smaller impact parameters many more shells could be intersected and need to be included.